### PR TITLE
[Salesforce] Alternative authentication flow

### PR DESCRIPTION
**Pull request by nick-Ag**
This PR adds authentication using username & password to Salesforce. This was requested by a customer and the functionality will be locked to that customer. See the ticket for more info: https://segment.atlassian.net/browse/STRATCONN-3682

Prior to creating the Salesforce client we will request an authentication token from Salesforce using the username + password if they are defined in the settings. This token will be used for all requests made from the request client.

Related PR: https://github.com/segmentio/app/pull/20349

## Testing

Tested successfully in staging, the username+password flow is used correctly and events are delivered.

In event tester the access token request using username + password can be seen by the user:
Screenshot 2024-05-02 at 11 29 35 AM

Subsequent requests using that token succeed:
Screenshot 2024-05-02 at 11 29 50 AM

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [x] [Segmenters] Tested in the staging environment

### CODE REVIEW

File: packages/destination-actions/src/destinations/salesforce/sf-operations.ts [hunk 1]
```diff
@@ -27,6 +28,72 @@ const validateSOQLOperator = (operator: string | undefined): SOQLOperator => {
...
+  const passwordRequestClient = createRequestClient({
+    headers: {
+      Authorization: `Bearer ${accessToken}`
+    }
+  })
+
+  return passwordRequestClient
+}
+
+const constructPassword = (password: string, securityToken?: string): string => {
```
Comment (marinhero) [inline packages/destination-actions/src/destinations/salesforce/sf-operations.ts:52]: Please add a comment here to indicate when a combination is required and when it isn't

**Review by marinhero — CHANGES_REQUESTED** [review]

File: packages/destination-actions/src/destinations/salesforce/sf-operations.ts [hunk 2]
```diff
@@ -27,6 +28,72 @@ const validateSOQLOperator = (operator: string | undefined): SOQLOperator => {
...
+  return combined
+}
+
+export const authenticateWithPassword = async (
+  username: string,
+  auth_password: string,
+  security_token?: string,
+  isSandbox?: boolean
+): Promise<RefreshAccessTokenResult> => {
+  const clientId = process.env.SALESFORCE_CLIENT_ID
```
[the remark this answers is not in the document]
Comment (nick-Ag) [reply packages/destination-actions/src/destinations/salesforce/sf-operations.ts:71]: These have always existed in prod

**Review by marinhero — APPROVED** [review]

**Review by marinhero — CHANGES_REQUESTED** [review]
Revoking approve until the flagon issue is fixed

**Review by marinhero — APPROVED** [review]

**Comment by nick-Ag** [comment]
@varadarajan-tw or @joe-ayoub-segment To whoever deploys this, I will handle the `push` operation when I'm ready to release this

**Comment by joe-ayoub-segment** [comment]
@nick-Ag said this is OK to deploy

**Comment by joe-ayoub-segment** [comment]
hi @nick-Ag PR deployed! Not pushed as requested.

**Pull request merged**